benchmark/datasets.py

```diff
@@ -369,6 +384,75 @@ def __init__(self, nb_M=1000):
     def distance(self):
         return "euclidean"
 
+class BigANNDatasetT3(DatasetCompetitionFormat):
+    def __init__(self, nb_M=1000):
+        self.nb_M = nb_M
+        self.nb = 10**6 * nb_M
+        self.d = 128
+        self.nq = 10000
+        self.dtype = "uint8"
+        self.ds_fn = "base.1B.u8bin"
+        self.qs_fn = "query.public.10K.u8bin"
+        self.gt_fn = "GT.public.1B.ibin" if self.nb == 10**9 else None
+        self.base_url = "https://dl.fbaipublicfiles.com/billion-scale-ann-benchmarks/bigann/"
+        self.basedir = os.path.join(BASEDIR, "bigann")
+
+    def get_dataset_fn(self):
+        ds_fn = super().get_dataset_fn()
+        ds_fn_fp = ds_fn + ".float32"
+        if os.path.exists( ds_fn_fp ):
+            return ds_fn_fp
+        else:
+            return ds_fn
+
+    def get_dataset(self):
+        ds_fn = super().get_dataset_fn()
+        ds_fn_fp = ds_fn + ".float32"
+        if os.path.exists( ds_fn_fp ):
+            return xbin_mmap_floatarray( ds_fn_fp, ( self.nb, self.d ) )
+        else:
+            return super().get_dataset()
+
+    def get_groundtruth(self,k=None):
+        if self.gt_fn:
+            return super().get_groundtruth(k)
+        else: # assume its a crop
+            gt_crop = os.path.join( self.basedir, "gt.ibin.crop_nb_%d" % self.nb )
+            if os.path.exists(gt_crop):
+                gt = xbin_mmap(os.path.join( gt_crop), dtype=np.int32)
+                if k is not None:
+                    assert k <= 100
+                    gt = gt[:, :k]
+                return gt
+            else:
+                raise Exception("No ground truth file.")
+
+    def get_queries(self):
+        ds_fn = super().get_dataset_fn()
+        ds_fn_fp = ds_fn + ".float32"
+        if os.path.exists( ds_fn_fp ):
+            filename = os.path.join(self.basedir, self.qs_fn)
+            return xbin_mmap_with_cast( filename, np.uint8, np.float32 )
+        else:
+            return super().get_queries()
+
+    def get_dataset_iterator(self, bs=512, split=(1,0)):
+        ds_fn = super().get_dataset_fn()
+        ds_fn_fp = ds_fn + ".float32"
+        if os.path.exists( ds_fn_fp ):
+            nsplit, rank = split
+            i0, i1 = self.nb * rank // nsplit, self.nb * (rank + 1) // nsplit
+            filename = self.get_dataset_fn()
+            x = xbin_mmap_floatarray(filename, (self.nb, 128))
+            assert x.shape == (self.nb, self.d)
+            for j0 in range(i0, i1, bs):
+                j1 = min(j0 + bs, i1)
+                yield sanitize(x[j0:j1])
+        else:
+            return super().get_dataset_iterator(bs, split)
+
+    def distance(self):
+        return "euclidean"
```

Comment:
Why is this necessary? Is this for debugging?

Reply:
Yeah.  It was for debugging.  Removing.